Docs update (#73)
solidity-labs-io/forge-proposal-simulator: docs/guides/governor-bravo-proposal.md

    ```solidity
    function build()
        public
        override
        buildModifier(addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO"))
    {
        /// STATICCALL -- non-mutative and hence not recorded for the run stage

        // Get the vault address
        address bravoVault = addresses.getAddress("BRAVO_VAULT");

        // Get the token address
        address token = addresses.getAddress("BRAVO_VAULT_TOKEN");

        // Get the timelock bravo's token balance
        uint256 balance = Token(token).balanceOf(
            addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO")
        );

        /// CALLS -- mutative and recorded

        // Whitelist the deployed token on the deployed vault
        Vault(bravoVault).whitelistToken(token, true);

        // Approve the token for the vault
        Token(token).approve(bravoVault, balance);

        // Deposit all tokens into the vault
        Vault(bravoVault).deposit(token, balance);
    }
    ```

-   `deploy()`: Deploys any necessary contracts. This example demonstrates the deployment of Vault and an ERC20 token. Once deployed, these contracts are added to the `Addresses` contract by calling `addAddress()`.

    ```solidity

>     function deploy() public override {

Reviewer: order of how we display steps to users is important and should follow this pattern throughout all documentation:

1. name
2. description
3. deploy
4. afterDeployMock (if applicable)
5. build
6. run
7. simulate
8. validate

Author: done

        vm.makePersistent(address(addresses));

        // Set governor bravo. This address is used for proposal simulation and check on
        // chain proposal state.
        setGovernor(addresses.getAddress("PROTOCOL_GOVERNOR"));

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

## Proposal Simulation

Before proceeding with the tutorial, ensure all addresses in `Addresses.json` are removed. There are two methods for executing proposals:

1. **Using `forge test`**: Detailed information on this method can be found in the [integration-tests.md](../testing/integration-tests.md) section.
2. **Using `forge script`**: This tutorial employs this method.

### Setting Up Your Deployer Address

The deployer address is used to broadcast the transactions deploying the proposal contracts. Ensure your deployer address holds sufficient funds from the faucet to cover deployment costs on the testnet. We prioritize security in private key management. To avoid storing the private key as an environment variable, we utilize Foundry's cast tool. Ensure the cast address matches the Deployer address.

> If you lack a wallet in `~/.foundry/keystores/`, create one by executing:

Reviewer: ```suggestion
If there are no wallets in the `~/.foundry/keystores/` folder, create one by executing:
```

Author: Done

        assertEq(amount, balance);

        // Ensure all minted tokens are deposited into the vault
        assertEq(token.balanceOf(address(bravoVault)), token.totalSupply());
    }
    ```

## Proposal Simulation

Before proceeding with the tutorial, ensure all addresses in `Addresses.json` are removed. There are two methods for executing proposals:

1. **Using `forge test`**: Detailed information on this method can be found in the [integration-tests.md](../testing/integration-tests.md) section.
2. **Using `forge script`**: This tutorial employs this method.

### Setting Up Your Deployer Address

The deployer address is used to broadcast the transactions deploying the proposal contracts. Ensure your deployer address holds sufficient funds from the faucet to cover deployment costs on the testnet. We prioritize security in private key management. To avoid storing the private key as an environment variable, we utilize Foundry's cast tool. Ensure the cast address matches the Deployer address.

If there are no wallets in the `~/.foundry/keystores/` folder, create one by executing:

```sh
cast wallet import ${wallet_name} --interactive
```

> 

Reviewer 2: We repeat this section in all the guides and mainnet examples, so let's move to the introduction section and remove from all the other places.

Author: done.

            addresses.addAddress("BRAVO_VAULT", address(bravoVault), true);
            bravoVault.transferOwnership(owner);
        }

        // Deploy the token address if not already deployed, transfer ownership to the timelock
        // and transfer all initial minted tokens from the deployer to the timelock
        if (!addresses.isAddressSet("BRAVO_VAULT_TOKEN")) {
            Token token = new Token();
            addresses.addAddress("BRAVO_VAULT_TOKEN", address(token), true);
            token.transferOwnership(owner);

            // During forge script execution, the deployer of the contracts is
            // the DEPLOYER_EOA. However, when running through forge test, the deployer of the contracts is this contract.
            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(address(owner), balance);
        }
    }
    ```

    Since these changes do not persist from runs themselves, after the contracts are deployed, the user must update the Addresses.json file with the newly deployed contract addresses.

> -   `build()`: Add actions to the proposal contract. [See the build function](../overview/architecture/proposal-functions.md#build-function). In this example, an ERC20 token is whitelisted on the Vault contract. Then the Governor bravo's timelock approves the token to be spent by the vault, and calls deposit on the vault. The actions should be written in solidity code and in the order they should be executed in the proposal. Any calls (except to the Addresses and Foundry Vm contract) will be recorded and stored as actions to execute in the run function. The `caller` address that will call actions is passed into `buildModifier`; it is the Governor bravo's timelock for this example. The `buildModifier` is a necessary modifier for the `build` function and will not work without it.

Reviewer: ```suggestion
-   `build()`: Add actions to the proposal contract. [See the build function](../overview/architecture/proposal-functions.md#build-function). In this example, an ERC20 token is whitelisted on the Vault contract. Then the Governor Bravo's timelock approves the token to be spent by the vault, and calls deposit on the vault. The actions should be written in solidity code and in the order they should be executed in the proposal. Any calls (except to the Addresses and Foundry Vm contract) will be recorded and stored as actions to execute in the run function. The `caller` address that will call actions is passed into `buildModifier`; it is the Governor Bravo's timelock for this example. The `buildModifier` is a necessary modifier for the `build` function and will not work without it.
```

    ```solidity
    function build()
        public
        override
        buildModifier(addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO"))
    {
        /// STATICCALL -- non-mutative and hence not recorded for the run stage

        // Get the vault address
        address bravoVault = addresses.getAddress("BRAVO_VAULT");

        // Get the token address
        address token = addresses.getAddress("BRAVO_VAULT_TOKEN");

        // Get the timelock bravo's token balance
        uint256 balance = Token(token).balanceOf(
            addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO")
        );

        /// CALLS -- mutative and recorded

        // Whitelist the deployed token on the deployed vault
        Vault(bravoVault).whitelistToken(token, true);

        // Approve the token for the vault
        Token(token).approve(bravoVault, balance);

        // Deposit all tokens into the vault
        Vault(bravoVault).deposit(token, balance);
    }
    ```

> -   `run()`: Sets up the environment for running the proposal. [See the run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId`, and `governor`, and then calls `super.run()` to run the entire proposal. In this example, `primaryForkId` is set to `sepolia` for executing the proposal. Next, the `addresses` object is set by reading from the `addresses.json` file. The Governor bravo address to simulate the proposal through is set using `setGovernor`. This will be used to check onchain calldata and simulate the proposal.

Reviewer: ```suggestion
-   `run()`: Sets up the environment for running the proposal. [See the run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId`, and `governor`, and then calls `super.run()` to run the entire proposal. In this example, `primaryForkId` is set to `sepolia` for executing the proposal. Next, the `addresses` object is set by reading from the `addresses.json` file. The Governor Bravo address to simulate the proposal through is set using `setGovernor`. This will be used to check onchain calldata and simulate the proposal.
```

    ```solidity
    function run() public override {
        // Create and select sepolia fork for proposal execution.
        primaryForkId = vm.createFork("sepolia");
        vm.selectFork(primaryForkId);

        // Set addresses object reading addresses from json file.
        setAddresses(
            new Addresses(
                vm.envOr("ADDRESSES_PATH", string("addresses/Addresses.json"))
            )
        );

        // Set governor bravo. This address is used for proposal simulation and check on
        // chain proposal state.
        setGovernor(addresses.getAddress("PROTOCOL_GOVERNOR"));

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

> -   `simulate()`: For governor bravo proposal, this function is defined in the governance specific contract and needs not to be overridden. This function executes the proposal actions outlined in the `build()` step. First required number of governance tokens are minted to the proposer address. Proposer delegates votes to himself and then proposes the proposal. Then the time is skipped by the voting delay, proposer casts vote and the proposal is queued. Next, time is skipped by the timelock delay and then finally the proposal is executed. Check the code snippet below with inline comments to get a better idea.

Reviewer: ```suggestion
-   `simulate()`: For Governor Bravo proposals, this function is defined in the governance specific contract and needs not be overridden. This function executes the proposal actions outlined in the `build()` step. The following steps are run when simulating a proposal. First, the required number of governance tokens are minted to the proposer address. Second, the proposer delegates votes to himself and proposes the proposal. Then the time is skipped by the voting delay, proposer casts votes and the proposal is queued. Next, time is skipped by the timelock delay and then finally, the proposal is executed in the timelock. View the code snippet below with inline comments for an example.
```

After adding FPS into project dependencies, the next step is the creation of the
first Proposal contract. This example provides guidance on writing a proposal
for deploying new instances of `Vault.sol` and `Token`. These contracts are

> located in the fps-example-repo [mocks](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/mocks). Copy each file used in this tutorial into your project for running examples or clone [fps-example-repo](https://github.com/solidity-labs-io/fps-example-repo/) repo and you have everything setup there. For this tutorial it's assumed that you have cloned the fps-example-repo.

Reviewer: ```suggestion
located in the fps-example-repo [mocks](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/mocks). Copy each file in this tutorial into your project, or clone the [fps-example-repo](https://github.com/solidity-labs-io/fps-example-repo/) repo. This tutorial assumes you have cloned the fps-example-repo.
```

            bravoVault.transferOwnership(owner);
        }

        if (!addresses.isAddressSet("BRAVO_VAULT_TOKEN")) {
            Token token = new Token();
            addresses.addAddress("BRAVO_VAULT_TOKEN", address(token), true);
            token.transferOwnership(owner);

            // During forge script execution, the deployer of the contracts is
            // the DEPLOYER_EOA. However, when running through forge test, the deployer of the contracts is this contract.
            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(address(owner), balance);
        }
    }

    function build()
        public
        override
        buildModifier(addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO"))
    {
        /// STATICCALL -- not recorded for the run stage
        address bravoVault = addresses.getAddress("BRAVO_VAULT");
        address token = addresses.getAddress("BRAVO_VAULT_TOKEN");
        uint256 balance = Token(token).balanceOf(
            addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO")
        );

>         Vault(bravoVault).whitelistToken(token, true);

Reviewer: this should be under the

`        /// CALLS -- mutative and recorded`

section

Author: done

-   `name()`: Define the name of your proposal.
-   `description()`: Provide a detailed description of your proposal.

> -   `run()`: Sets environment for running the proposal. It sets `addresses`, `primaryForkId` and `governor`. `addresses` is address object

Reviewer: ```suggestion
-   `run()`: Sets environment for running the proposal. It sets `addresses`, `primaryForkId` and `governor`. `addresses` in address object
```

This section is not clearly elucidated, a reader is going to read this and get lost.

We need to lay this out on a silver platter, and assume the person who is going to pick this up doesn't really know anything about how to use this tool.

Author: added code snippet with inline comments and also added reference to `proposal-functions.md` file.

Reviewer: this section still isn't ready to go because the run function is at the start, and we wanted to organize the functions by which way makes most sense to execute them

Author: Done

>     containing addresses to be used in proposal that are fetched from `Addresses.json`. `primaryForkId` is the RPC URL or alias of the blockchain that will be used to simulate the proposal actions and broadcast if any contract deployment is required.`governor` is the address of the governor bravo contract.

Reviewer: this is so unclearly worded.

Author: added code snippet with inline comments and also added reference to `proposal-functions.md` file.

-   `deploy()`: Deploy any necessary contracts. This example demonstrates the deployment of Vault and an ERC20 token. Once the contracts
    are deployed, they are added to the `Addresses` contract by calling `addAddress()`.
-   `build()`: Set the necessary actions for your proposal. In this example,
    ERC20 token is whitelisted on the Vault contract. The actions should be
    written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. `caller` address that will call actions is passed into `buildModifier`, it is the governor bravo's timelock for this example.

> -   `validate()`: This final step is crucial for validating the post-execution state. It

Reviewer: ```suggestion
-   `validate()`: This final step validates the system in its post-execution state. It
```

>     ensures that the governor bravo's timelock is the new owner of Vault and token, the tokens were transferred to governor bravo's timelock and the token was whitelisted on the Vault contract

Reviewer: this whole section might make more sense if at each section you show the code and explain what each line of the function does.

Author: Have added code snippets with inline comments for each function.

## Proposal simulation

> First of all, please remove all addresses in `Addresses.json` before running through the tutorial. Now it's time to proceed with execution. There are two options available:

Reviewer: ```suggestion
First of all, remove all addresses in `Addresses.json` before running through the tutorial. To proceed with execution, there are two options available:
```

1. **Using `forge test`**: Details on this method can be found in the [integration-tests.md](../testing/integration-tests.md) section.
2. **Using `forge script`**: This is the chosen method for this tutorial.

## Running the Proposal with `forge script`

### Setting Up Your Deployer Address

> The deployer address is the the one you'll use to broadcast the transactions deploying the proposal contracts. Ensure your deployer address has enough funds from the faucet to cover deployment costs on the testnet.

Reviewer: ```suggestion
The deployer address is the one used to broadcast the transactions deploying the proposal contracts. Ensure your deployer address has enough funds from the faucet to cover deployment costs on the testnet.
```

anything else on how this process works and you have to use the forge cli command to enter the private key @anajuliabit

> We prioritize security when it comes to private key management. To avoid storing the private key as an environment variable, we use Foundry's cast tool.

Reviewer: I'd add something here about the address stored in cast should match the deployer EOA. These are two separate bullet points and it makes sense to have them be a single one.

Author: done.

You'll need a Bravo Governor contract set up on the testnet before running the proposal.

We have a script in [script/](https://github.com/solidity-labs-io/fps-example-repo/tree/main/script) folder called `DeployGovernorBravo.s.sol` to facilitate this process.

Before running the script, you must add the `DEPLOYER_EOA` address to the `Addresses.json` file.

```json
[
    {
        "addr": "YOUR_DEV_ADDRESS",
        "name": "DEPLOYER_EOA",
        "chainId": 11155111,
        "isContract": false
    }
]
```

> After adding the addresses, run the script:

Reviewer: ```suggestion
After adding the address, run the script:
```

### Setting Up the Addresses JSON

The last step before running the proposal is to add the DEPLOYER_EOA address

> to Addresses.json. The final Addresses.json file should be something like this:

Reviewer: didn't we just cover this? why are we repeating the same steps we just said previously

Author: updated

```

If a password was provide to the wallet, the script will prompt for the password before broadcasting the proposal.

> A DAO member can check whether the calldata proposed on the governance matches the calldata from the script exeuction. It is important to note that two new addresses have been added to the `Addresses.sol` storage. These addresses are not included in the JSON file and must be added manually for accuracy.

Reviewer: ```suggestion
A DAO member can check whether the calldata proposed on the governance matches the calldata from the script exeuction. It is important to note that two new addresses have been added to the `Addresses.sol` storage. These addresses are not included in the JSON file and must be added manually as new contracts have now been added to the system.
```

        "chainId": 11155111,
        "isContract": true,

>         "name": "PROTOCOL_GOVERNANCE_TOKEN"

Reviewer: naming convention: throughout the examples to make things perfectly clear that a user is supposed to fill in their address maybe do "0x< YOUR_GOVERNANCE_TOKEN_ADDRESS >" so it's really clear they have to replace it with their own value

forge script script/InitializeBravo.s.sol --rpc-url sepolia --broadcast -vvvv --slow --sender ${wallet_address} -vvvv --account ${wallet_name} -g 200
```

### Setting Up the Addresses JSON

> Copy the `GOVERNOR_BRAVO_ALPHA` address from the script output and add it to the `Addresses.json` file. The final `Addresses.json` file should be something like this:

Reviewer: ```suggestion
Copy the `GOVERNOR_BRAVO_ALPHA` address from the script output and add it to the `Addresses.json` file. The final `Addresses.json` file should follow this structure:
```

```

A DAO member can verify whether the calldata proposed on the governance matches the calldata from the script execution. It's crucial to note that two new addresses have been added to the `Addresses.sol` storage. These addresses are not included in the JSON file and must be added manually as new contracts have now been added to the system.

> The proposal script will deploy the contracts in the `deploy()` method and will generate actions calldata for each individual action along with proposal calldata for the proposal. The proposal can be manually proposed using the cast send along with the calldata generated above.

Reviewer: ```suggestion
The proposal script will deploy the contracts in the `deploy()` method and will generate actions calldata for each individual action along with proposal calldata for the proposal. The proposal can be manually proposed using `cast send` along with the calldata generated above.
```
